### PR TITLE
feat(webpack-builder): remove addDefaultPlugins & add new module `shared/plugins.ts`

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Description

Waiting for #1681 ...

- Remove `addDefaultPlugins`.
- Add new module `src/shared/plugins.ts`.
  - re-exports plugins from `src/plugins/*`.
  - refactor the loading and executing of plugins.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by chenjiahan** [comment]
这个 PR 需要 resolve 一下冲突哈

File: packages/builder/webpack-builder/src/shared/plugin.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  assetsRetry: 'PluginAssetsRetry',
+} as const;
+
+export type BuiltinPluginExport = typeof builtinPluginExports;
+export type BuiltinPluginName = keyof BuiltinPluginExport;
+
+export const loadBuiltinPlugin = async (plugin: BuiltinPluginName) => {
+  const exportName = builtinPluginExports[plugin];
+  const module = await import(`../plugins/${plugin}`);
+  return module[exportName] as () => BuilderPlugin;
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/shared/plugin.ts:52]: 这样写的话，就丢失了插件的类型了，如果以后某个插件需要接受一个入参，会有类型问题。

Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/plugin.ts]: https://github.com/modern-js-dev/modern.js/pull/1703/commits/7f2656e18f3e5c698df0a5336a53212c02fb7797#diff-4d4724b2f8195b706fb2c2d15f8b6faedf36dba4dcf82d1aed0b9a464ef6c3b3R44-R95

@chenjiahan 可以通过这样的类型映射让 `loadBuiltinPlugin` 返回详细的类型，不过 `loadBuiltinPlugins` 好像没办法获取带有详细类型的元组。入参的话考虑修改 `loadBuiltinPlugins` 允许接受函数的话可以写成这样：

```typescript
export const presetMinimal = [
  'basic',
  'entry',
  'cache',
  'target',
  'output',
  devtool => devtool(),
  'resolve',
] as const;

export const applyMinimalPlugins = () => applyBuiltinPlugins(presetMinimal);
```

或者改回类似 `addDefaultPlugins` 那样手动书写逻辑会更好嘛

File: packages/builder/webpack-builder/src/shared/plugin.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  'basic',
+  'entry',
+  'cache',
+  'target',
+  'output',
+  'devtool',
+  'resolve',
+] as const;
+
+export const applyMinimalPlugins = () => applyBuiltinPlugins(presetMinimal);
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/shared/plugin.ts:117]: 这里真的需要一个 map 来处理吗，感觉把逻辑变得复杂了，直接这样写是不是也可以

```js
export const applyMinimalPlugins = () => Promise.all([
  import('../plugins/a'),
  import('../plugins/b'),
  ...
]);
```

Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/plugin.ts:117]: 好嘞

Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/plugin.ts]: https://github.com/modern-js-dev/modern.js/blob/239acc834318dc52629520532c6c489fcf3b4f7b/packages/builder/webpack-builder/src/shared/plugin.ts#L7-L28

@chenjiahan 改成这样了

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
🎉

**Pull request merged**